eclass/ruby-ng.eclass

```diff
@@ -356,21 +368,40 @@ _RUBY_IMPLEMENTATIONS_DEPEND=""
 ruby_implementations_depend() {
 	debug-print-function ${FUNCNAME} "${@}"
 
+	_ruby_set_globals_invalidate_if_stale
 	_ruby_implementations_depend
 	echo "${_RUBY_IMPLEMENTATIONS_DEPEND}"
 }
 
 _ruby_implementations_depend() {
+	_ruby_set_globals_invalidate_if_stale
+
 	local depend _ruby_implementation
 	for _ruby_implementation in "${_RUBY_GET_ALL_IMPLS[@]}"; do
 		depend="${depend}${depend+ }ruby_targets_${_ruby_implementation}? ( $(ruby_implementation_depend $_ruby_implementation) )"
 	done
 	_RUBY_IMPLEMENTATIONS_DEPEND="${depend}"
 }
 
-_ruby_get_all_impls
-_ruby_get_use_targets
-_ruby_implementations_depend
+_ruby_set_globals() {
+	_RUBY_SET_GLOBALS_USE_RUBY="${USE_RUBY}"
+	_ruby_get_all_impls
+	_ruby_get_use_targets
+	_ruby_implementations_depend
+}
+
+_ruby_set_globals_invalidate_if_stale() {
+	# Packages may try to restrict their test dependencies to ease bootstrapping/porting
+	# if they're not yet available for a newer Ruby implementation by setting
+	# USE_RUBY="<some subset of original USE_RUBY>" ruby_add_bdepend ...
+	if [[ ${_RUBY_SET_GLOBALS_USE_RUBY} != ${USE_RUBY} && -z ${_RUBY_SET_GLOBALS_INVALIDATING} ]] ; then
+		_RUBY_SET_GLOBALS_INVALIDATING=1
```

Comment:
Wouldn't `local` work here?

Reply:
yes, good point

Reply:
wait, do you mean `local` or `local -x`? We need it to be visible in any subcalls to avoid recursion

Reply:
`local` is fine. It only means it doesn't affect use after function leaves, it gets propagated to calls normally.

```
a() {
	echo $FOO
}

b() {
	local FOO=bar
	a
}

b
```

Reply:
Oh, thank you!